src/mbgl/map/vector_tile_data.cpp

```diff
@@ -44,34 +44,62 @@ void VectorTileData::parse() {
         const VectorTile* vt = &vectorTile;
         TileParser parser(*vt, *this, style, glyphAtlas, glyphStore, spriteAtlas, sprite);
         parser.parse();
+
+        if (state == State::obsolete) {
+            return;
+        }
+
+        if (parser.isPartialParse()) {
+            state = State::partial;
+        } else {
+            state = State::parsed;
+        }
     } catch (const std::exception& ex) {
         Log::Error(Event::ParseTile, "Parsing [%d/%d/%d] failed: %s", id.z, id.x, id.y, ex.what());
         state = State::obsolete;
         return;
     }
-
-    if (state != State::obsolete) {
-        state = State::parsed;
-    }
 }
 
 void VectorTileData::render(Painter &painter, const StyleLayer &layer_desc, const mat4 &matrix) {
-    if (state == State::parsed && layer_desc.bucket) {
-        auto databucket_it = buckets.find(layer_desc.bucket->name);
-        if (databucket_it != buckets.end()) {
-            assert(databucket_it->second);
-            databucket_it->second->render(painter, layer_desc, id, matrix);
-        }
+    if (!ready() || !layer_desc.bucket) {
+        return;
+    }
+
+    Bucket* bucket = getBucket(layer_desc.bucket->name);
+    if (!bucket) {
+        return;
     }
+
+    bucket->render(painter, layer_desc, id, matrix);
 }
 
 bool VectorTileData::hasData(const StyleLayer &layer_desc) const {
-    if (state == State::parsed && layer_desc.bucket) {
-        auto databucket_it = buckets.find(layer_desc.bucket->name);
-        if (databucket_it != buckets.end()) {
-            assert(databucket_it->second);
-            return databucket_it->second->hasData();
-        }
+    if (!ready() || !layer_desc.bucket) {
+        return false;
+    }
+
+    Bucket* bucket = getBucket(layer_desc.bucket->name);
+    if (!bucket) {
+        return false;
     }
-    return false;
+
+    return bucket->hasData();
+}
+
+Bucket* VectorTileData::getBucket(const std::string& key) const {
+    std::lock_guard<std::mutex> lock(bucketsMutex);
+
+    auto it = buckets.find(key);
+    if (it == buckets.end())
+        return nullptr;
+
+    assert(it->second);
+    return it->second.get();
```

Comment:
This may fail when `setBucket` overwrites an existing bucket. In that case, the existing bucket will be deallocated and a new one will be created, leaving the returned pointer invalid. `setBucket` needs to ensure that it will never overwrite an existing bucket.